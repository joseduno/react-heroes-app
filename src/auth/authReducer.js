import { types } from "../types/types"

export const authReducer = (state={}, action) => {



    switch (action?.type) {
        
        case types.login:
            return {
                // No entiendo por que utilizar el operador spread
                ...action.payload,
                logged: true
            }

        case types.logout:
            return {
                logged: false
            }
    
        default:
            return state
        }
}