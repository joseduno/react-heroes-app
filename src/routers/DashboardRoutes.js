import React from 'react'
import { Switch, Route, Redirect } from 'react-router-dom'
import { DcSreen } from '../components/dc/DcSreen'
import { HeroScreen } from '../components/heroes/HeroScreen'
import { MarvelScreen } from '../components/marvel/MarvelScreen'
import { SearchScreen } from '../components/search/SearchScreen'
import { Navbar } from '../components/ui/Navbar'

export const DashboardRoutes = () => {
    return (
        <>
        {/* Los Routers hijos no tienen el componente BrowserRouter as Routes */}
            <Navbar />
            <div className='container mt-3'>
                <Switch>
                    <Route exact path='/marvel' component={MarvelScreen} />
                    <Route exact path='/dc' component={DcSreen} />
                    <Route exact path='/hero/:heroId' component={HeroScreen} />
                    <Route exact path='/search' component={SearchScreen} />

                    <Redirect to='/marvel' />

                </Switch>
            </div>
        </>
    )
}
