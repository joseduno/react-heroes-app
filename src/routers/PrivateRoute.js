import React from 'react'
import { Redirect, Route } from 'react-router-dom'
import PropTypes from 'prop-types';

/* Meconfunde un poco el use de ...rest y props */

export const PrivateRoute = ({
    isAuthenticated,
    component: Component,
    ...rest
}) => {

    localStorage.setItem('lastPath', rest.location.pathname)

    return (
        /* Al utilizar un router secundario, no utilizar la propiedad 'exact' */
        <Route {...rest}
            component={props => (
                (isAuthenticated)
                ? <Component {...props} />
                : <Redirect to='/login' />
            )}
        />
    )
}


PrivateRoute.protoTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    component: PropTypes.func.isRequired,
}