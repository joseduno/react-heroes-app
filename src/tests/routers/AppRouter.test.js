import { mount } from "enzyme"
import { AuthContext } from "../../auth/AuthContext"
import { AppRouter } from "../../routers/AppRouter"

describe('Pruebas en el AppRouter', () => {

    
    test('debe mostrar el login si el usuario no esta autenticado', () => {

        const contextValue = {
            dispatch: jest.fn(),
            user: {logged: false}
        }

        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <AppRouter />
            </AuthContext.Provider>
        )

        expect(wrapper).toMatchSnapshot()
    })

    test('debe mostrar el componente de /marvel si estÃ¡ autenticado', () => {

        const contextValue = {
            dispatch: jest.fn(),
            user: {logged: true, name: 'test'}
        }

        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <AppRouter />
            </AuthContext.Provider>
        )
        expect(wrapper.find('.navbar').exists()).toBe(true)
    })
    
    
})
