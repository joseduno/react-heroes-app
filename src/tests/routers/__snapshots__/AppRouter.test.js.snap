// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el AppRouter debe mostrar el login si el usuario no esta autenticado 1`] = `
<div>
  <div
    className="container mt-5"
  >
    <h1>
      Login
    </h1>
    <hr />
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Ingresar
    </button>
  </div>
</div>
`;
